### PR TITLE
[fix]: [CDS-86605]: Fix calculating parent-type and finding the node in schema for given FQN in YAML

### PR DESCRIPTION
**Pull request by brijeshdhakar**
## Describe your changes

Fix calculating parent-type and finding the node in schema for given FQN in YAML.
Change:

- Fix calculating the parent_type for a input in YAML: Earlier we used to iterate the root jsonNode for the given FQN and the last type field was considered as the parent_type. That is not true.
- Now we traverse the FQN and find the index of last FQNNode that has NODE_TYPE as UUID and its parent is in the list of possible parentOfGroupNodes. And node at that path would be the root node. And we return the type of that node now.

- Fix finding the schema for a field: Earlier we used to recursively traverse the schema node for any type and find the target field. Also if some field value is in terms of $ref then any field inside that reference field would not be found by current logic.
- Now we use the fqnFromParentNode and traverse at exactly the required path. And also handles the $ref everytime so all the child fields would be found now.

Testing:

- Executed the old unit-test for finding the parentType. Worked fine. Added one more cases that was failing with earlier code but pass with new logic.
- Tested e2e the input-schema api with fields at different levels and the input-schema metadata is coming correcty. The YAML that i tested is below:
```
version: 1
kind: pipeline
spec:
  inputs:
    duration:
      type: string
      description: timeout of shell script step
      required: true
    script:
      type: string
      default: ls
    jenkinsJobName:
      type: string
    jenkinsConnector:
      type: string
  stages:
    - type: deployment
      id: deploy_1
      timeout: 10m
      spec:
        environment: 
          environmentRef: Env
        service:
          serviceRef: Svc
        steps:
          - type: k8s-rolling-deploy
            name: k8s_rolling
            timeout: 10m
            spec:
              skipDryRun: false
              pruningEnabled: false
          - type: JenkinsBuild
            name: jenkins_build_step
            spec:
              connectorRef: <+inputs.jenkinsConnector>
              jobName: <+inputs.jenkinsJobName>
      failure: 
        errors: 
          - all
        action: 
          type: abort
    - type: custom
      name: s1
      identifier: fdas
      id: cs_1
      spec:
        steps:
          - type: shell-script
            name: shell_script_1
            id: shell_script_1
            identifier: shell_script_1
            spec:
              shell: bash
              on_delegate: true
              source:
                type: inline
                spec:
                  script: <+inputs.script>
            timeout: <+inputs.duration>
      exports:
        export_1:
          value: <+stage.spec.steps.shell_script_1.type>
          desc: "Some description"
```

And the response has the correct metadata and depdendencies:
```
{
    "inputs": [
        {
            "details": {
                "name": "duration",
                "type": "string",
                "description": null,
                "required": true,
                "execution": false,
                "allowed_values": null,
                "regex": null,
                "default": null
            },
            "metadata": {
                "field_properties": [],
                "dependencies": {
                    "required_runtime_inputs": [],
                    "required_fixed_values": []
                }
            }
        },
        {
            "details": {
                "name": "script",
                "type": "string",
                "description": null,
                "required": false,
                "execution": false,
                "allowed_values": null,
                "regex": null,
                "default": "ls"
            },
            "metadata": {
                "field_properties": [],
                "dependencies": {
                    "required_runtime_inputs": [],
                    "required_fixed_values": []
                }
            }
        },
        {
            "details": {
                "name": "jenkinsJobName",
                "type": "string",
                "description": null,
                "required": false,
                "execution": false,
                "allowed_values": null,
                "regex": null,
                "default": null
            },
            "metadata": {
                "field_properties": [
                    {
                        "input_type": "string",
                        "internal_type": "jenkins_job_name"
                    }
                ],
                "dependencies": {
                    "required_runtime_inputs": [
                        {
                            "field_name": "connectorRef",
                            "input_name": "jenkinsConnector"
                        }
                    ],
                    "required_fixed_values": []
                }
            }
        },
        {
            "details": {
                "name": "jenkinsConnector",
                "type": "string",
                "description": null,
                "required": false,
                "execution": false,
                "allowed_values": null,
                "regex": null,
                "default": null
            },
            "metadata": {
                "field_properties": [
                    {
                        "input_type": "string",
                        "internal_type": "jenkins_connector"
                    }
                ],
                "dependencies": {
                    "required_runtime_inputs": [],
                    "required_fixed_values": []
                }
            }
        }
    ]
}
```

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56967)

### CODE REVIEW

**Comment by brijeshdhakar** [comment]
trigger uts

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 10 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56967)* status: 12 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/yaml/schema/inputs/InputsSchemaServiceImpl.java` line 74 at r1](https://reviewable.io/reviews/harness/harness-core/56967#-Nm4eJOx9H1mpfUji-Ru:-Nm4eJOx9H1mpfUji-Rv:bo5qmxu) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/953-yaml-commons/src/main/java/io/harness/yaml/schema/inputs/InputsSchemaServiceImpl.java#L74)):*
> ```Java
>           // fetch corresponding input-details for the given template yaml
>           String parentNodeType =
>               YamlInputUtils.getParentNodeTypeForGivenFQNField(new YamlNode(yamlConfig.getYamlMap()), fqn);
> ```

can we optimise this?



_Code quote:_
```Java
YamlInputUtils.getParentNodeTypeForGivenFQNField(new YamlNode(yamlConfig.getYamlMap()), fqn);
```
___
*[`template-service/service/src/main/java/io/harness/yaml/individualschema/BaseTemplateSchemaParser.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/56967#-Nm4eqgj4oCyZ70efKWd:-Nm4eqgj4oCyZ70efKWe:b-jfepef) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/template-service/service/src/main/java/io/harness/yaml/individualschema/BaseTemplateSchemaParser.java#L63)):*
> ```Java
>     String fqnFromParentNode = inputFieldMetadata.getFqnFromParentNode();
>     // TODO Refactor: Use TemplateSchemaMetadata and create TemplateSchemaRequest instead of using
>     // PipelineSchemaRequest.
> ```

@brijeshdhakar can you create ticket for this

File: 953-yaml-commons/src/main/java/io/harness/yaml/individualschema/AbstractStaticSchemaParser.java [hunk 1]
```diff
@@ -155,22 +159,63 @@ JsonNode getReferredJsonNodeByPath(String path) {
...
     }
-    // if the input-field is directly in the node yaml(parallel to spec) then parentFieldNode properties should have
-    // that field. Else find the field in the spec.
-    if (parentFieldNode.has(PROPERTIES_NODE) && parentFieldNode.get(PROPERTIES_NODE).has(fieldName)) {
-      return parentFieldNode.get(PROPERTIES_NODE).get(fieldName);
+    return getFieldByFQN(parentFieldNode, Arrays.asList(individualSchemaRequest.getFqnFromParentNode().split("\\.")), 0,
+        MAX_DEPTH_FOR_RECURSION);
+  }
+
+  private JsonNode getFieldByFQN(JsonNode jsonNode, List<String> fqn, int currentIndex, int depth) {
```
Comment (utkarshchoubey-harness) [inline 953-yaml-commons/src/main/java/io/harness/yaml/individualschema/AbstractStaticSchemaParser.java:173]: can we add a desciption of what this function is doing

File: 953-yaml-commons/src/main/java/io/harness/yaml/individualschema/InputFieldMetadata.java [hunk 2]
```diff
@@ -25,7 +25,7 @@ public class InputFieldMetadata {
   String parentNodeType; // eg: JiraCreate, custom etc.
   FQN fqn;
 
-  private static List<String> baseNodeGroups = Arrays.asList("stage", "stages", "step", "steps");
+  public static List<String> parentTypesOfNodeGroups = Arrays.asList("stages", "steps");
```
Comment (utkarshchoubey-harness) [inline 953-yaml-commons/src/main/java/io/harness/yaml/individualschema/InputFieldMetadata.java:28]: is the same for both v0 and v1 ?

**Comment by brijeshdhakar** [comment]
*[`953-yaml-commons/src/main/java/io/harness/yaml/schema/inputs/InputsSchemaServiceImpl.java` line 74 at r1](https://reviewable.io/reviews/harness/harness-core/56967#-Nm4eJOx9H1mpfUji-Ru:-Nm4r_lt7h-8DrjqT2lH:bsfgwio) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/953-yaml-commons/src/main/java/io/harness/yaml/schema/inputs/InputsSchemaServiceImpl.java#L74)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can we optimise this?
</blockquote></details>

I don't see any scope of optimisation here. The existing utils work on yamlNode only. But we can pass the jsonNode from here and convert into YamlNode in the YamlInputUtils::getParentNodeTypeForGivenFQNField for better reading.

**Comment by brijeshdhakar** [comment]
*[`template-service/service/src/main/java/io/harness/yaml/individualschema/BaseTemplateSchemaParser.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/56967#-Nm4eqgj4oCyZ70efKWd:-Nm4ruYeDohSntQV4NUU:b-rx4pgr) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/template-service/service/src/main/java/io/harness/yaml/individualschema/BaseTemplateSchemaParser.java#L63)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

@brijeshdhakar can you create ticket for this
</blockquote></details>

Created the ticket: https://harness.atlassian.net/browse/CDS-87227

**Comment by brijeshdhakar** [comment]
*[`template-service/service/src/main/java/io/harness/yaml/individualschema/BaseTemplateSchemaParser.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/56967#-Nm4eqgj4oCyZ70efKWd:-Nm4ryUxD31C_-RsBGNB:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/template-service/service/src/main/java/io/harness/yaml/individualschema/BaseTemplateSchemaParser.java#L63)):*
<details><summary><i>Previously, brijeshdhakar (Brijesh Dhakar) wrote…</i></summary><blockquote>

Created the ticket: https://harness.atlassian.net/browse/CDS-87227
</blockquote></details>

Done.

**Comment by brijeshdhakar** [comment]
*[`953-yaml-commons/src/main/java/io/harness/yaml/individualschema/InputFieldMetadata.java` line 28 at r1](https://reviewable.io/reviews/harness/harness-core/56967#gh-1432273275:-Nm57KZt6j0swcPtv7y-:b3wd8xd) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/953-yaml-commons/src/main/java/io/harness/yaml/individualschema/InputFieldMetadata.java#L28)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

is the same for both v0 and v1 ?
</blockquote></details>

Yes, This list is the parent of nodeGroups. And we start our FQN from the next path of these.

**Comment by brijeshdhakar** [comment]
trigger uts

**Review by brijeshdhakar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56967)* status: 12 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/yaml/individualschema/AbstractStaticSchemaParser.java` line 173 at r1](https://reviewable.io/reviews/harness/harness-core/56967#gh-1432271881:-Nm5G14Z1R3TqOf35nF4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/953-yaml-commons/src/main/java/io/harness/yaml/individualschema/AbstractStaticSchemaParser.java#L173)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

can we add a desciption of what this function is doing
</blockquote></details>

Done.
___
*[`953-yaml-commons/src/main/java/io/harness/yaml/schema/inputs/InputsSchemaServiceImpl.java` line 74 at r1](https://reviewable.io/reviews/harness/harness-core/56967#-Nm4eJOx9H1mpfUji-Ru:-Nm5G2pK9si-ylNHvxN5:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/71fd4338ef66e61b568ada52e1928367da229fe9/953-yaml-commons/src/main/java/io/harness/yaml/schema/inputs/InputsSchemaServiceImpl.java#L74)):*
<details><summary><i>Previously, brijeshdhakar (Brijesh Dhakar) wrote…</i></summary><blockquote>

I don't see any scope of optimisation here. The existing utils work on yamlNode only. But we can pass the jsonNode from here and convert into YamlNode in the YamlInputUtils::getParentNodeTypeForGivenFQNField for better reading.
</blockquote></details>

Done

**Comment by brijeshdhakar** [comment]
trigger uts

**Comment by brijeshdhakar** [comment]
trigger uts

**Review by utkarshchoubey-harness — APPROVED** [review]
lgtm!

**Comment by brijeshdhakar** [comment]
trigger ut7 ut8 ut9

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 12 files at r1, 3 of 4 files at r2, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56967)* status: 12 file(s) to review. 2. unresolved discussions.

**Comment by brijeshdhakar** [comment]
trigger ut8 ut9

**Comment by brijeshdhakar** [comment]
trigger ut7 ut9 review

**Comment by brijeshdhakar** [comment]
trigger ut8

**Pull request merged**